### PR TITLE
Move DBReader off of pwomise and onto async/await

### PR DESCRIPTION
**Pull request by itsjunetime**
I don't trust `Pwomise` and I was getting a lot of weird hangs while working on stuff today, so I converted all of the `DBReader` stuff onto async/await and I think it feels much nicer. It also forces handling of errors when calling async stuff (while previously, with `Promise`s, you could just ignore the errors) so maybe that'll catch some previous errors and weirdness we would run into.

Also ensure IM logs everything to stdout upon barcelona startup with the `_IMLogForceWriteToStdout`. I'm not like 100% certain that imessage is now logging to stdout, but theoretically it is and I can't figure out why it wouldn't. Also this function is theoretically already called by our `imagentd`, but I worry that it could be overwritten by another process in between running imagentd upon startup and running barcelona (plus we want this to happen during our own debugging), so I think it's still a good move.

@myyra what do you think? I think this is a good move, maybe you could do a quick look through (or run it on your machine) and let me know if anything seems off?

### CODE REVIEW

File: Core/Barcelona/Extensions/Glue/Publishers.swift [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+            for item in self {
+                group.addTask {
+                    await transform(item)
+                }
+            }
+
+            return await group.reduce(into: []) { $0.append($1) }
+        }
+    }
+}
```
Comment (myyra) [inline Core/Barcelona/Extensions/Glue/Publishers.swift]: Are these part of this PR? I can't see them used anywhere.

Additionally, I'm not really a fan of the magic retaining sink. Creating a `var cancellables = Set<AnyCancellable>()` ensures predictable lifetimes, auto cancellation on `deinit` etc., we should really favor those.

File: Core/Barcelona/Queries/IMDPersistence.swift [hunk 2]
```diff
@@ -256,31 +287,38 @@ public func BLLoadChatItems(withGraph graph: [String: ([String], IMServiceStyle)
...
+        if let chat = chats.first(where: { $0.id == value.chatID }) {
+            // Then, if it's new to the dictionary, we just insert it
+            if dict[chat.id] == nil {
+                dict[chat.id] = ([value.messageID], chat.service)
+            } else {
+                // Else, we append it to what's already there
+                // And we have to do the nasty `.0.0` thing because subscript can return a (K, V) tuple,
+                // and swift is inferring that's what we wnat here, so we have to grab the value from the
+                // tuple that it returns, then append to the first item in that tuple.
+                dict[chat.id]?.0.append(value.messageID)
```
Comment (myyra) [inline Core/Barcelona/Queries/IMDPersistence.swift:316]: Really good commenting 👌

**Review by myyra — APPROVED** [review]
This is really awesome! A very welcome change for overall codebase quality as well.

I was planning to move to GRDB 6, so we can get concurrency support in the database functions themselves, this will be a great help.

I left one comment regarding some Combine stuff that didn't seem to be a part of this PR, but to me, this is good to go already.

File: Core/Barcelona/Extensions/Glue/Publishers.swift (hunk 1 above)
Comment (itsjunetime) [reply Core/Barcelona/Extensions/Glue/Publishers.swift]: Oh yeah, I was using these during development but I'm not certain all of them actually ended up in the PR. I'll fix that.

And I used to be a fan of using a set, but I feel like it's not applicable for all circumstances. For example, a lot of functions in barcelona don't have an associated class (they're just global functions) and thus can't utilize a set of cancellables. Plus, if you use a set for cancellables that aren't supposed to be long-running (e.g. a single read from the database, even though we're not using combine for that), you can fill up the set with cancellables that aren't wanted anymore and cause a small memory leak. I think the set pattern should only be used in the case of pipelines/cancellables that should stay alive during the entire lifetime of the object.

**Comment by itsjunetime** [comment]
I'm going to merge this in via CLI so that I can deal with the conflicts more manually

File: Core/Barcelona/Extensions/Glue/Publishers.swift (hunk 1 above)
Comment (myyra) [reply Core/Barcelona/Extensions/Glue/Publishers.swift]: Good points 👍 I'd still favor anything explicit, though, maybe with properties if there's a concern for filling up the set (which is valid in some cases and one I never understood about Combine's design). For global functions, just forwarding the pipeline enough for something to save it properly would be the best, but really I don't feel like demanding anything "best" with this codebase right now.

But let's see how to approach those the next time something comes up. This is probably a meaningless discussion anyway, as I feel concurrency is a much better fit here for many reasons.

**Pull request merged**